### ISSUE TITLE
Errors compiling

### ISSUE THREAD

**Issue by BenWoodford** [swlines/CIFReader#1]
I installed all the dependencies, compiled MySQL++ appropriately and I'm greeted with this when I try to compile CIFReader

> in file included from classes/NRCIF.h:25:0,
>                  from classes/NRCIF.cpp:22,
>                  from main.cpp:21:
> classes/sqlNetRail.h:24:14: error: expected constructor, destructor, or type conversion before ‘(’ token
> In file included from main.cpp:21:0:
> classes/NRCIF.cpp: In static member function ‘static bool NRCIF::processFile(mysqlpp::Connection&, const char_)’:
> classes/NRCIF.cpp:108:10: error: ‘schedules_t’ was not declared in this scope
> classes/NRCIF.cpp:108:21: error: template argument 1 is invalid
> classes/NRCIF.cpp:108:21: error: template argument 2 is invalid
> classes/NRCIF.cpp:108:37: error: invalid type in declaration before ‘;’ token
> In file included from main.cpp:21:0:
> classes/NRCIF.cpp:330:97: error: ‘class CIFRecordNRBS’ has no member named ‘train_uid’
> classes/NRCIF.cpp:374:21: error: expected ‘;’ before ‘row’
> classes/NRCIF.cpp:409:22: error: ‘row’ was not declared in this scope
> classes/NRCIF.cpp:412:21: error: expected ‘;’ before ‘row’
> classes/NRCIF.cpp:447:22: error: ‘row’ was not declared in this scope
> In file included from main.cpp:21:0:
> classes/NRCIF.cpp:108:23: warning: unused variable ‘scheduleInsert’ [-Wunused-variable]
> classes/NRCIF.cpp: In static member function ‘static void NRCIF::runSchedules(mysqlpp::Connection&, std::vector<locations_t>&, std::vector<locations_change_t>&, std::vector<int>&)’:
> classes/NRCIF.cpp:655:2: error: ‘SizeThresholdInsertPolicy’ is not a member of ‘mysqlpp::Query’
> classes/NRCIF.cpp:655:44: error: expected primary-expression before ‘>’ token
> classes/NRCIF.cpp:655:64: error: ‘insert_policy’ was not declared in this scope
> classes/NRCIF.cpp:656:8: error: ‘class mysqlpp::Query’ has no member named ‘insertfrom’
> classes/NRCIF.cpp:659:8: error: ‘class mysqlpp::Query’ has no member named ‘insertfrom’
> classes/NRCIF.cpp: In static member function ‘static void NRCIF::runSchedulesStpCancel(mysqlpp::Connection&, std::vector<CIFRecordNRBS_>&, std::vector<CIFRecordNRBS*>&)’:
> classes/NRCIF.cpp:704:91: error: ‘class CIFRecordNRBS’ has no member named ‘train_uid’
> classes/NRCIF.cpp:725:2: error: ‘SizeThresholdInsertPolicy’ is not a member of ‘mysqlpp::Query’
> classes/NRCIF.cpp:725:44: error: expected primary-expression before ‘>’ token
> classes/NRCIF.cpp:725:64: error: ‘insert_policy’ was not declared in this scope
> classes/NRCIF.cpp:726:8: error: ‘class mysqlpp::Query’ has no member named ‘insertfrom’
> main.cpp: In function ‘int main(int, char**)’:
> main.cpp:229:39: error: ‘class boost::filesystem3::directory_entry’ has no member named ‘leaf’

**Comment by swlines** [comment]
Am looking into this. I will be committing a new version of the code which should fix most of these problems in the next few days.

**Comment by BenWoodford** [comment]
Discovered that the first issue with the sqlNetRail.h file may be due to me using 3.0.9, as 3.1.0 would not compile for me. I will try with 3.1.0 again and see if I still have this issue.

tl;dr first error was me being an idiot.

**Comment by BenWoodford** [comment]
Did MySQL++ 3.1.0 throw link errors when you built it?

**Comment by swlines** [comment]
Not that I recall, no. I've built MySQL++ 3.1.0 on both OS X 10.7.2 and Debian Squeeze 6.0.2.

**Comment by BenWoodford** [comment]
What version of GCC? Getting errors with 4. :S

**Comment by swlines** [comment]
On OS X:

$ gcc -v
Using built-in specs.
Target: i686-apple-darwin11
Configured with: /private/var/tmp/llvmgcc42/llvmgcc42-2336.1~1/src/configure --disable-checking --enable-werror --prefix=/Developer/usr/llvm-gcc-4.2 --mandir=/share/man --enable-languages=c,objc,c++,obj-c++ --program-prefix=llvm- --program-transform-name=/^[cg][^.-]*$/s/$/-4.2/ --with-slibdir=/usr/lib --build=i686-apple-darwin11 --enable-llvm=/private/var/tmp/llvmgcc42/llvmgcc42-2336.1~1/dst-llvmCore/Developer/usr/local --program-prefix=i686-apple-darwin11- --host=x86_64-apple-darwin11 --target=i686-apple-darwin11 --with-gxx-include-dir=/usr/include/c++/4.2.1
Thread model: posix
gcc version 4.2.1 (Based on Apple Inc. build 5658) (LLVM build 2336.1.00)

On Debian:

tom@charon:~$ gcc -v
Using built-in specs.
Target: x86_64-linux-gnu
Configured with: ../src/configure -v --with-pkgversion='Debian 4.4.5-8' --with-bugurl=file:///usr/share/doc/gcc-4.4/README.Bugs --enable-languages=c,c++,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.4 --enable-shared --enable-multiarch --enable-linker-build-id --with-system-zlib --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.4 --libdir=/usr/lib --enable-nls --enable-clocale=gnu --enable-libstdcxx-debug --enable-objc-gc --with-arch-32=i586 --with-tune=generic --enable-checking=release --build=x86_64-linux-gnu --host=x86_64-linux-gnu --target=x86_64-linux-gnu
Thread model: posix
gcc version 4.4.5 (Debian 4.4.5-8)

**Comment by BenWoodford** [comment]
Much obliged, MySQL 3.1.0 doesn't compile with 4.6 but will with 4.4.

I'm just going to keep posting my resolutions here so that anyone on updated repos (I'm on Natty so everything's fairly up to date) can resolve their issues.

**Comment by swlines** [comment]
No prob. 

Will get the new version committed once I can roll out some bugs in the version I'm currently playing with.

**Comment by BenWoodford** [comment]
> classes/NRCIF.cpp: In static member function ‘static bool NRCIF::processFile(mysqlpp::Connection&, const char_)’:
> classes/NRCIF.cpp:330:97: error: ‘class CIFRecordNRBS’ has no member named ‘train_uid’
> In file included from main.cpp:21:0:
> classes/NRCIF.cpp: In static member function ‘static void NRCIF::runSchedulesStpCancel(mysqlpp::Connection&, std::vector<CIFRecordNRBS_>&, std::vector<CIFRecordNRBS*>&)’:
> classes/NRCIF.cpp:704:91: error: ‘class CIFRecordNRBS’ has no member named ‘train_uid’
> main.cpp: In function ‘int main(int, char**)’:
> main.cpp:229:39: error: ‘class boost::filesystem3::directory_entry’ has no member named ‘leaf’

The first error about train_uid can be fixed by altering the two mentions of scheduleDetail->train_uid in NRCIF.cpp to scheduleDetail->uid. Chances are this was a typo?

**Comment by swlines** [comment]
Judging by the code I'm currently working on, yes, definitely a typo.

**Comment by swlines** [comment]
To fix the main.cpp issue (I can't commit this as it has some database wrangling in it too...)...replace the obvious bits with this:

```
bool operationFailed = false;

for(vector<string>::iterator fit = files.begin(); fit < files.end(); ++fit) {
    /* hack to get around weird seg fault */
    mysqlpp::Connection connection;
    if(!connection.connect(sqlDatabase, sqlServer, sqlUsername, sqlPassword)) {
        cerr << "DB connection failed:" << conn.error() << endl;
        return 1;
    }

    connection.set_option(new mysqlpp::MultiStatementsOption(true));

    if(networkRailCIF){
        if(!NRCIF::processFile(connection, (*fit).c_str())) {
            operationFailed = true;
            connection.disconnect();
            break;
        }
    }
    else {}

    connection.disconnect();
}

files.clear();

if(!operationFailed) {
    for(vector<string>::iterator dit = directories.begin(); dit < directories.end(); ++dit) {
        boost::filesystem::directory_iterator end_itr;
        string filePath = *dit;
        if(filePath.substr(filePath.length() -1, 1) != "/") filePath += "/";

        set<string> filePaths;

        for(boost::filesystem::directory_iterator itr(*dit); itr != end_itr; ++itr) {
            if(boost::filesystem::is_directory(*itr)) continue;
            else if(boost::filesystem::exists(*itr)) {
                filePaths.insert(filePath + itr->path().filename().c_str());
            }
        }

        for(set<string>::iterator it = filePaths.begin(); it != filePaths.end(); it++) {
            /* hack to get around weird seg fault */
            mysqlpp::Connection connection;
            if(!connection.connect(sqlDatabase, sqlServer, sqlUsername, sqlPassword)) {
                cerr << "DB connection failed:" << conn.error() << endl;
                return 1;
            }

            connection.set_option(new mysqlpp::MultiStatementsOption(true));

            if(networkRailCIF){
                if(!NRCIF::processFile(connection, (*it).c_str())) {
                    operationFailed = true;
                    connection.disconnect();
                    break;
                }
            }

            connection.disconnect();
        }

        filePaths.clear();
    }
}

directories.clear();
```

In addition, the new code should fix any issues with ATOCs CIF files due to them not being compliant with the Network Rail spec. _sigh_

**Comment by BenWoodford** [comment]
I'm probably doing something basic wrong so I won't make a new issue:

> Performing run on Network Rail CIF: data/TTISF618.MCA
> Importing DFTTISS created for CFCATE
> Generated on 2012-02-03 at 2113
> Data from 2012-02-03 to 2012-09-30
> 
> # File type: FULL;  file size; 444152 kilobytes
> 
> 0%   10   20   30   40   50   60   70   80   90   100%
> |----|----|----|----|----|----|----|----|----|----|
> *ERROR: Unable to locate association to STP cancel (main UID W00971, assoc UID W00973)
> ERROR: Failure has taken place. Update will not be completed...

What did I break? :<

**Comment by swlines** [comment]
Ah, no, this is an actual issue. This surrounds the fact that the CIF files (prefixed .MCA) that ATOC supply are not actually compliant with the Network Rail format. This is fixed (again) in the next commit.

**Comment by BenWoodford** [comment]
Public Transport groups not adhering to standards and keeping up to date? Such a new concept. :(

Ah well, it's building now so I'm happy, I'll just wait for the next build and start playing with it.

**Comment by swlines** [comment]
Indeed :(

The sad thing is more the fact that they claim it meets the spec (which it does, almost)... 

The problem is that Network Rail says that trains should be expected where schedules with an STP indicator of C must come after a P. i.e. you can't cancel a schedule before it exists...which is what ATOC try to do.

**Comment by swlines** [comment]
Just doing a few last tests and then will commit the latest stuff up.

There are a few things which aren't the same in this version (for various reasons) and the database schema has changed. Will all be detailed in the CHANGES file.

**Comment by swlines** [comment]
ATOC parsing and boost filesystem issues fixed in 8252bf4
